src/rgw/rgw_auth.h

```diff
@@ -303,6 +303,7 @@ class TokenExtractor {
 public:
   virtual ~TokenExtractor() = default;
   virtual std::string get_token(const req_state* s) const = 0;
+  virtual std::string get_service_token(const req_state* s) const = 0;
```

Comment:
instead of adding a `get_service_token()` call to all `TokenExtractor`s, i think that `rgw::auth::swift::DefaultStrategy` should just have two kinds of `TokenExtractor` - one that returns the existing `HTTP_X_AUTH_TOKEN`, and a new one that returns `HTTP_X_SERVICE_TOKEN`. that way, other auth strategies like `rgw::auth::sts::DefaultStrategy` don't have to implement a do-nothing `get_service_token()`

something like this?
```diff
--- a/src/rgw/rgw_swift_auth.h
+++ b/src/rgw/rgw_swift_auth.h
@@ -181,7 +181,6 @@ public:
 
 
 class DefaultStrategy : public rgw::auth::Strategy,
-                        public rgw::auth::TokenExtractor,
                         public rgw::auth::RemoteApplier::Factory,
                         public rgw::auth::LocalApplier::Factory,
                         public rgw::auth::swift::TempURLApplier::Factory {
@@ -200,12 +199,19 @@ class DefaultStrategy : public rgw::auth::Strategy,
   using aplptr_t = rgw::auth::IdentityApplier::aplptr_t;
   using acl_strategy_t = rgw::auth::RemoteApplier::acl_strategy_t;
 
-  /* The method implements TokenExtractor for X-Auth-Token present in req_state. */
-  std::string get_token(const req_state* const s) const override {
-    /* Returning a reference here would end in GCC complaining about a reference
-     * to temporary. */
-    return s->info.env->get("HTTP_X_AUTH_TOKEN", "");
-  }
+  /* The implements TokenExtractor for X-Auth-Token present in req_state. */
+  struct AuthTokenExtractor : rgw::auth::TokenExtractor {
+    std::string get_token(const req_state* const s) const override {
+      return s->info.env->get("HTTP_X_AUTH_TOKEN", "");
+    }
+  } auth_token_extractor;
+
+  /* The implements TokenExtractor for X-Service-Token present in req_state. */
+  struct ServiceTokenExtractor : rgw::auth::TokenExtractor {
+    std::string get_token(const req_state* const s) const override {
+      return s->info.env->get("HTTP_X_SERVICE_TOKEN", "");
+    }
+  } service_token_extractor;
 
   aplptr_t create_apl_remote(CephContext* const cct,
                              const req_state* const s,
```

Reply:
I will investigate this.

Reply:
Much cleaner, thanks! I've included the proposal in the PR.